pkgs/_macro_host/lib/macro_host.dart

```diff
@@ -15,31 +15,23 @@ import 'package:macro_service/macro_service.dart';
 /// Tools that want to support macros, such as the Analyzer and the CFE, can
 /// do so by running a `MacroHost` and providing their own `HostService`.
 class MacroHost implements HostService {
-  final MacroServer macroServer;
-  final ListOfServices services;
+  late final MacroServer _macroServer;
+  final QueryService queryService;
   final MacroBuilder macroBuilder = MacroBuilder();
   final MacroRunner macroRunner = MacroRunner();
 
   // TODO(davidmorgan): this should be per macro, as part of tracking per-macro
   // lifecycle state.
   Completer<Set<int>>? _macroPhases;
 
-  MacroHost._(this.macroServer, this.services) {
-    services.services.insert(0, this);
-  }
+  MacroHost._(this.queryService);
 
-  /// Starts a macro host serving the provided [service].
-  ///
-  /// The service passed in should handle introspection RPCs, it does not need
-  /// to handle others.
-  ///
-  /// TODO(davidmorgan): make this split clearer, it should be in the protocol
-  /// definition somewhere which requests the host handles.
-  static Future<MacroHost> serve({required HostService service}) async {
-    final listOfServices = ListOfServices();
-    listOfServices.services.add(service);
-    final server = await MacroServer.serve(service: listOfServices);
-    return MacroHost._(server, listOfServices);
+  /// Starts a macro host with introspection queries handled by [queryService].
+  static Future<MacroHost> serve({required QueryService queryService}) async {
+    final result = MacroHost._(queryService);
+    final server = await MacroServer.serve(service: result);
+    result._macroServer = server;
```

Comment:
This is definitely still sending up red flags for me, assigning manually fields like this after construction is something I would always try to avoid. I get that the cyclic dependency sort of requires it here, but I would still like to see that cycle broken.

Reply:
The circle is due to the class doing too much: it's the host and the host service.

Split out the host service and now it looks much better, thanks :)